boards: phytec: Add phyBOARD-Electra AM64x

dnltz opened the pull request:
Add support for TI's AM6442 SOC, which is almost identical to the AM62x. Moreover, it adds support for the phyBOARD-Electra kit, which includes the phyCORE-AM64x SOM.

This PR uses the same board names as introduced in #70913

dnltz committed:
soc: ti: k3: Add support for AM6442

The AM64x and AM62x are both SOCs from the TI K3 family
and share common architecture designs. The M4F subsystem
is actuall identical on both SOCs.

Therefore, just add all missing CONFIGs, files, etc. to
support the AM6442x SOC.

Since MMR and RAT initialization are identical too, both
functions can be re-used. However, since they might
differ in the future, the am64x has it's own init
function.

Signed-off-by: Daniel Schultz <d.schultz@phytec.de>

dnltz committed:
dts: arm: ti: am62x_m4: Use DT_FREQ_x fpr frequenies

We define two frequencies in the am62x_m4.dtsi file.

Use DT_FREQ_M for both frequency to make them more
human-readable and easier to understand.

Signed-off-by: Daniel Schultz <d.schultz@phytec.de>

dnltz committed:
dts: arm: ti: Add dtsi for AM64x M4

This file is basically a copy of the AM62x M4 dtsi but an
additional mcu_uart1 interface.

The internal clock frequency feeded into the UART IP is
96 MHz instead of 48 MHz, which is different to the AM62x.

Signed-off-by: Daniel Schultz <d.schultz@phytec.de>

dnltz committed:
board: phytec: Add phyBOARD-Electra

Add PHYTEC's phyBOARD-Electra kit with the phyCORE-AM64x
System on Module (SOM).

This board only supports the M4F core, UART and GPIO right now,
since other drivers are still missing.

Signed-off-by: Daniel Schultz <d.schultz@phytec.de>

dnltz commented on dts/arm/ti/am64x_m4.dtsi:
```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright (c) 2023 Texas Instruments Incorporated
```
This is an almost identical copy of the am62x_m4. Should I keep this Copyright or change it?

vaishnavachath replied:
@dnltz There is no need to retain the copyright in this particular case. However we need some strategy to not duplicate these SoC level dtsi files (atleast for different cores in same SoC), in this case it is okay.

nordicjm commented on boards/phytec/phyboard_electra/Kconfig.phyboard_electra:
```diff
@@ -0,0 +1,10 @@
+
+config BOARD_PHYBOARD_ELECTRA
+	select SOC_AM6442_M4 if BOARD_PHYBOARD_ELECTRA_AM6442_M4
+	help
+	  PHYTEC phyBOARD-Electra AM64x M4
```
```suggestion
```

nordicjm commented on boards/phytec/phyboard_electra/doc/img/phyCORE-AM64x_Electra_frontside.webp:
Reduce filesize using https://tinypng.com/

dnltz replied:
the original file size is 1.5MB and this link only saves 135kB. I already resized the original PNG to 30% before converting it to wepb. I think 100kB are allowed for pictures.

vaishnavachath commented on boards/phytec/phyboard_electra/doc/index.rst:
```diff
@@ -0,0 +1,143 @@
+********
+
+The Linux running on the A53 uses the remoteproc framework to manage the M4F co-processor.
+Therefore, the testing requires the binary to be copied to the SD card to allow the A53 cores to
```
It would be nice to provide instructions to do the same from u-boot console, that provides a faster development environment. I do something like below for development on m4/a53 on AM62 SK
```
dhcp; setenv serverip 192.168.0.108; rproc init; rproc stop 0; tftp
0x82000000 zephyr.elf; rproc load 0 0x82000000 $filesize;  rproc start
0;tftp 0x82000000 zephyr.bin; dcache flush; icache flush; dcache off;
icache off; go 0x82000000
```

dnltz replied:
Thanks for sharing those commands with me! We currently figuring out how to improve Zephyr/MCU+ integration into our products and they will be very helpful.

I will update documentation for both products later. Kinda thinking this might be a nice uboot environment function...

vaishnavachath commented on soc/ti/k3/am6x/m4/soc.c:
```diff
@@ -71,3 +72,16 @@ static int am62x_m4_init(void)
+{
+	sys_mm_drv_ti_rat_init(
+		region_config, ADDR_TRANSLATE_RAT_BASE_ADDR, ARRAY_SIZE(region_config));
+	mmr_unlock();
```
What was the intent to duplicate this?  I see you are using the exact same RAT configs for AM62 and AM64, I don't think that is a good idea even though basic functionality works fine, there should be some flexibility to configure this per SoC with different memory map as one mapping configuration might not work everywhere.

dnltz replied:
I'm not totally sure both architectures are 100% identical.

I thought about renaming the am62 init to `am6x_m4_init`. Guess I will do that...

vaishnavachath replied:
@dnltz sorry I think my comment was confusing, what I meant was that the intent to duplicate should be also to use separate RAT configurations, having the same is not a good idea for long term when more and more support is added.

gramsay0 commented on soc/ti/k3/am6x/m4/soc.c:
```diff
@@ -44,30 +52,88 @@ static struct address_trans_region_config region_config[] = {
 SYS_INIT(am62x_m4_init, EARLY, CONFIG_KERNEL_INIT_PRIORITY_DEFAULT);
+#endif
+
+#ifdef CONFIG_SOC_AM6442_M4
```
If this is same for AM6234/AM6442, I'd just wrap in in `CONFIG_SOC_SERIES_AM6X_M4` rather than duplicate, there are lots of SoCs in this family

dnltz replied:
@vaishnavachath what do you think? Maybe we should introduce SOC_AM62X_M4/SOC_AM64x_M4 configs? I don't think we need to define socs with the full part number like AM6234 here in Zephyr since we don't care about them.

gramsay0 replied:
From a quick look on the TI site there is currently AM62/AM64/AM65/AM67/AM68/AM69 series.
IMHO creating duplication later is easy, removing it is difficult 🤷

vaishnavachath replied:
@gramsay0  @dnltz I checked the TI SDK offerings, the default configurations are same for AM64 and AM62 so it can be reused now without duplication. 
But for the rest of the devices AM67/AM68/AM69 .etc even though the devices are of same family with increasing capabilities, the memory map might be different and such reuse won't be possible, for example only some of the devices  offer a general purpose M4 core which is programmable by the user, the reuse and extension can apply to devices of same family - like AM62-AM62A-AM62P .etc

Another concern is that RAT configuration is use case dependent and a global RAT configuration might not be suitable for all use cases and in TI SDKs also the RAT configs are configurable per example/application. that  was one of the reasons suggesting duplicate configs so that when someone changes the default configs for one SoC family it should not cause silent failures in other SoC family, RAT config failures are a bit tricky to debug also since the driver does not expose any of the diagnostics features.

Reusing is good for now, but I will recommend atleast adding a comment mentioning that this configs are shared between multiple devices.

gramsay0 commented on dts/arm/ti/am64x_m4.dtsi:
```diff
@@ -0,0 +1,87 @@
+	gpio0: gpio@4201010 {
+		compatible = "ti,davinci-gpio";
+		reg = <0x4201010 0x100>;
+		#address-cells = <2>;
```
What does `#address-cells` do here?
I though `#gpio-cells` specified the number of cells for child bindings

dnltz replied:
Yeah that looks wrong. If I recall correctly, I added that because the build failed due missing address cells in earlier version. Anyhow, gonna remove that from both m4 dtsi files since it's working now.

gramsay0 commented on soc/ti/k3/am6x/CMakeLists.txt:
```diff
@@ -17,5 +17,16 @@ elseif(CONFIG_SOC_AM6234_M4)
   endif()
 
+  set(SOC_LINKER_SCRIPT ${CMAKE_CURRENT_SOURCE_DIR}/m4/linker.ld CACHE INTERNAL "")
+elseif(CONFIG_SOC_AM6442_M4)
```
Same here (duplication)

dnltz commented on boards/phytec/phyboard_electra/doc/index.rst:
```diff
@@ -0,0 +1,143 @@
+.. code-block:: console
+
+   # From the root of the Zephyr repository
+   west build -p -b phyboard_electra/am6442/m4 samples/hello_world
```
Thanks, that's definitely better.

dnltz committed:
dts: arm: ti: am62x_m4: Remove address-cells from gpio

An address-cells definition is not required for the gpio
node and therefore should be removed.

Signed-off-by: Daniel Schultz <d.schultz@phytec.de>

dnltz commented:
@gramsay0 had to rebase due a conflict in the MAINTAINERS.yml file.
